### ISSUE TITLE
JapanesePod doesn't find Kana words

### ISSUE THREAD

**Issue by bastik** [ospalh/anki-addons#45]
I have problems fetching the audio from JapanesePod for words that are usually written in Kana only. For example putting とても in the Kana field of the reading line (or in both the Kanji and Kana field), doesn't find the audio. If you look up the Kanji writing (which is 迚も), it works.

It would be nice, if  'download audio' could automate this as much as possible. An obvious source would be WWWJDIC, which has sort of an API. There are two modes: [raw](http://www.csse.monash.edu.au/~jwb/cgi-bin/wwwjdic.cgi?1ZUJ%E3%81%A8%E3%81%A6%E3%82%82) and [html](http://www.csse.monash.edu.au/~jwb/cgi-bin/wwwjdic.cgi?1MUJ%E3%81%A8%E3%81%A6%E3%82%82).
The advantage of the raw mode is that it is easy to parse, but the html mode has direct links to the audio URL if available.

Obviously one cannot continue when there are multiple words with the same reading, as they are sometimes pronounced differently.

I would have a go at this, but I'm unsure how to integrate it. Maybe first do the normal lookup and when it fails, try more fancy methods (without any additional GUI or notifications)?

**Comment by ospalh** [comment]
>(…)words that are usually written in Kana only(…)

Or rather, words written with *hira*gana, where there exist kanji versions. Katakana words work fine. I just tried 「アイスクリーム」,  other loanwords should work as well. They also have 「から」 and presumably a few other particles and pure hiragana words, like 「もしもし」. (But not 「そろそろ」. No idea why not.)

>(…) It would be nice, if 'download audio' could automate this as much as possible( by asking) WWWJDIC(…).

Well, i guess you *could* do that. But, honestly, i don’t feel like implementing that myself. You can currently try downloading the kanji-and-kana version by using the manual audio feature.

When you have any concrete questions about my code you can ask, but i think the code is reasonably clean and readable, so i guess you should first have a look at it yourself.

Also, i don’t know *why* they insist on, for example, 「貴方」 or 「貴女」, but then do *not* have 「貴男」? Or maybe you meant 「彼方」 instead.  Do you try 未だ or 未? And so on. 「４６４９」　almost makes more sense then writing it with 「宜]」.

**Comment by bastik** [comment]
In the patch, I do not attempt any kana normalization. This has the advantage that オス  is found, otherwise it would fail because there are multiple entries with the reading おす.
On the other hand, ゾウ fails, because the entry is missing. This seems to be an error in the dictionary as  this is spelling in wikipedia and other animals have their katakana included.

Also there is an encoding error when looking up サル. This is also a bug in wwwjdic, but I tested a lot of words and only had this problem once, so it shouldn't matter.

I took a conservative approach to cancel whenever there is a duplicate. This is a bit annoying as it happens quite a lot. One possible enhancement would be to accept duplicates if only one has the popular marker (P).

**Comment by ospalh** [comment]
Thanks for the code.
It looks interesting, an i will have a closer look at it.

There are a few things that, i, reading the code,  don’t understand. I’ll comment on some of them at the lines.

Also, i use the “[flow](http://nvie.com/posts/a-successful-git-branching-model/)” branching model. 
In short:
* The branch you do your changes on should be branched off `develop`, not `master`
* It should ideally be named feature-NN (feature-wwwjdic). This point is not too important
* The branch should then be merged back into develop. Please choose that one when you do 

So, could you maybe rebase your branch off develop, or branch off develop and cherry-pick your single commit? Or re-write the code. o_O I'll come to that.

The reason i talk about re-writing: Why didn’t you use BeautifulSoup? Look at some of the other [downloades](), use `Downloader.get_soup_from_url()`

Also, instead of expanding the JapanesPod downloader, use an extra WWWJDIC downloader and, maybe, only start that when the JapanesePod failed or returned blacklisted files.

**Comment by ospalh** [comment]
Btw, still doesn’t download anything for 「あなた」. Maybe that is the 
> conservative approach to cancel whenever there is a duplicate

Maybe you should just download the different versions if the input is ambiguous. As the review dialog shows the kanji the downloader got from WWWJDIC (quite nice!) it is up to the user to decide (or not care) which downloaded file is the right one (e.g. 貴方 vs. 彼方).

**Comment by ospalh** [comment]
On second thoughts, maybe but the WWWJDIC code in a separate downloader, and make that one download when the input consists of just hiragana. For the few cases where JapanesePod has pure kana words, you would get those files twice. I think that would be OK.

**Comment by bastik** [comment]
Thanks for the detailed feedback! A new version, that lists all the entries, is in preparation.
I'd like to show different kanji for each entry in downloads list, but the display text is currently an attribute of the downloader.

In `download.do_download` it takes for each `AudioDownloader` the `downloads_list` with triples `(temp_file_path, base_name, extras)` and converts this to 7-tuple list `retrieved_files_list`, which then looses the reference to the `AudioDownloader` object.

I guess to fix the problem, the `display_text` somehow has to be included in the `downloads_list` entries.

**Comment by bastik** [comment]
> On second thoughts, maybepbut the WWWJDIC code in a separate downloader, and make that one download when the input consists of just hiragana. For the few cases where JapanesePod has pure kana words, you would get those files twice. I think that would be OK.

I would prefer not to have a duplicates. Any idea how this could be achieved best with two separate downloader classes? For me functionality beats code aesthetics. ;-)

**Comment by ospalh** [comment]
>the display text is currently an attribute of the downloader

Oh right. I guess you could change *all* downloaders to make this by downloaded word. Sounds like too much work. Instead you can put it in the popup help, via that per word extras dictionary.

> I would prefer not to have a duplicates. 

When you use the downloader for English instead of for Japanese, you often get so many downloads that you pretty much have to skip some of them. So having a few (when you do it right very few) duplicates really doesn't bother me.

>For me functionality beats code aesthetics.

Surely they are limits, but your approaches all add lots of complexity. KISS

**Comment by bastik** [comment]
I'm determined to get it right, sorry... :)

>> the display text is currently an attribute of the downloader

> Oh right. I guess you could change all downloaders to make this by downloaded word. Sounds like too much work. Instead you can put it in the popup help, via that per word extras dictionary.

No problem, I can rework the downloaders. One could:

1. Extend triple to 4-tuple in `downloads_list`; remove attribute `display_text` from `downloader` class.
2. Create a new class `DownloadEntry` with all the properties that are currently in the triple + the remaining per entry properties like `display_text`, `show_skull_and_bones` and the icon stuff.

> So having a few (when you do it right very few) duplicates really doesn't bother me.

I agree that duplicates aren't that bad as they would be quite rare. But they can be easily avoided when all the code is in one class. After all, the audio files still come from japanesepod, it's just that the japanesepod website doesn't have a form where you can look up a hiragana word and get a list of all the audios with a word of that reading (at least not publicly available without login). wwwjdic offers exactly that: A convenient way to find japanesepod audios matching dictionary entries. So we use it as a lookup tool, but it is still essentially japanesepod.

**Comment by ospalh** [comment]
I am thinking about minimizing the programming effort, and coming up with a wwwjdic downloader would be good enough as it would give the pronunciations that are missing, and it would be somewhat easy to implement.

Actually reporting back the word we downloaded might be a good idea, though.

or maybe
 3. a [named tuple](https://docs.python.org/2.7/library/collections.html#collections.namedtuple).

> So we use (wwwjdic) as a lookup tool, but it is still essentially japanesepod.

That is borderline. The [Duden](https://github.com/ospalh/anki-addons/blob/master/downloadaudio/downloaders/duden.py) downloader gets its data from the Aussprachedatenbank der ARD. The difference there is that they host the files, but you do the lookup there.

(One minor point: when you make the JapanesePod downloader fetch these files, i think it would be a good idea to mention wwwjdic in the extras, or what you come up with to take its place.)

**Comment by bastik** [comment]
Named tuple sounds interesting, but I tried it with a new class. The advantage of a class is that it is nicer to add documentation, it allows to put default arguments in the constructor and it's also more standard and therefore easier to understand for casual python developers. :)

Please have a look if you are okay with that and I'll adapt the other downloaders as well:

https://github.com/bastik/anki-addons/compare/b4bf303caa777b...2e248d90b551af9b4

**Comment by ospalh** [comment]
Well, named tuples *are* *standard*, the `sys.version_info`, where Python keeps its own version, is one. The other points are mostly valid. I think you *can* use default values in named tuples. Advantages are that classes are easier to handle, you can set parameters one after another, and better known, so, yeah, a (this) class is fine.

**Comment by ospalh** [comment]
I’ve just had another idea:
Looks like the raw wwwjdic output would be easier to parse. How about just using that to “convert” the kana to kanji and then using the old style hit-and-hope lookup at JapanesePod?

**Comment by ospalh** [comment]
Please check out #48 before you spent too much (any) time on the wwwjdic part of your idea.

**Comment by bastik** [comment]
Code looks nice, but frankly I'm a bit surprised to see you chime in like that. First you say you are not interested in implementing it yourself and when I'm about finished, you can't be bothered to wait and look at the reworked patch.

Could you apply the branch, so I can merge my DownloadEntry code into that?

**Comment by ospalh** [comment]
> but frankly I'm a bit surprised to see you chime in like that

Oh, sorry. It’s that sometimes i like to kvetch a bit. Also, i still think of it as my project and when i think it is quicker to do something myself then wait for somebody else and then change that to my coding style, i’ll just write it myself.

**Comment by bastik** [comment]
See #49, pull request for the first part.

**Comment by bastik** [comment]
> Looks like the raw wwwjdic output would be easier to parse. How about just using that to “convert” the kana to kanji and then using the old style hit-and-hope lookup at JapanesePod?

I also thought about this. But it turns out, the html is that hard to parse (especially with BeautifulSoup it should be rather stable). I like to minimize the number of unnecessary requests to speed things up.

**Comment by ospalh** [comment]
Thanks again.
But that i pulled it doesn’t mean that i think the code is perfect.
When you put in 「あなた」 into my quick hack, you got back audio for 彼方, 貴方 and 貴女, with your code you only get one, 貴方. At least for 彼方 WWWJDIC has an [entry with a L](http://www.csse.monash.edu.au/~jwb/cgi-bin/wwwjdic.cgi?1MUJ%E3%81%82%E3%81%AA%E3%81%9F), that is, an audio link.　And that audio file is 彼方[あなた], *not*  彼方[かなた].

**Comment by bastik** [comment]
>  At least for 彼方 WWWJDIC has an entry with a L, that is, an audio link.　And that audio file is 彼方\[あなた\], not 彼方\[かなた\].

Hm, when clicking on the blue circle with the play sign in the center, I get the audio file for 彼方[かなた]. The html parser rejects this entry, because かなた doesn't match the requested あなた.

However, JapanesePod *does* have an audio for 彼方\[あなた\] ([link](http://assets.languagepod101.com/dictionary/japanese/audiomp3.php?kana=%E3%81%82%E3%81%AA%E3%81%9F&kanji=%E5%BD%BC%E6%96%B9)), which is distinct from the 貴方\[あなた\] audio. The reason for this is that there are multiple readings for 彼方, but WWWJDIC will link to at most one audio file! As it considers the かなた reading more common, this is the audio that is linked.

So the question is: Do we really need to find an audio for a reading that is not the most common? If so, we could use the hit-and-hope strategy by keeping the kanji in the languagepod download URL, but override the wrong kana with the kana that we want. (Shouldn't be too hard to change the code like this.)

Anyway, the case that there is more than one reading for a full kanji-fied word (not single kanji) is the exception and not that frequent. For the normal case (single reading) the html parser should work reliably and and help to avoid superfluous downloads of audio that we know doesn't exist. Keep in mind, that one not just gets 404, but a full download of this spoken error message for each failed attempt.

**Comment by ospalh** [comment]
I don't know about *need*. I think it’s nice if the downloaders fetch as many pronunciations as possible. So this new feature is nice as it is, as you get more pronunciation as before.

While i did change some of your downloader code (i think the definition should not be in the popup, and if so, at least the annotations like the part of speech (“(n)” for noun &c.) should be removed) i don’t think i will add code to fetch more meanings any time soon.

I think i will change the DownloadEntry class a bit. I have some ideas how having a class like that can simplify things some more. When you want to change the JapanesePod downloader to fetch some more, maybe you should wait a bit until i’m through with these changes, i.e. until the code doesn’t change every other day.

Anyway, i think i’ll publish this improved JapanesePod downloader soon.

**Comment by bastik** [comment]
> (i think the definition should not be in the popup, and if so, at least the annotations like the part of speech (“(n)” for noun &c.) should be removed)

There are different learning patterns for Japanese and it can be useful to first remember the reading of a word and at a later time add the kanji to your vocabulary. For example I know that 「つり」 means "fishing", but presented with the two kanji choices 釣り and 吊り, I would have no idea which one to pick. The translation in the popup saves you the lookup in an external dictionary.

> Anyway, i think i’ll publish this improved JapanesePod downloader soon.

Great! Thanks for accepting the patch and going through the trouble of fixing the bugs and coding style.

**Comment by ospalh** [comment]
Well, yeah, there usually are some Japanese words that i know that i you know the kanji for. But i have come to the realization a while ago that it is a good idea to add the kanji when creating the cards. You can then show just the kana with `{{kana:Reading}}`.
Actually i have an extra switch field where i put a single character (a ‘y', that is left from the Anki 1.2→2.0 converter) when i have learned all kanji of a word, and the card template goes from showing the reading to showing just the kanji. (And now i’m curious and have to look up 吊り…)

**Comment by ospalh** [comment]
That looks like bad timing.
WWWJDIC seems to have disappeared days before i was ready to roll this out, #56.

**Comment by bastik** [comment]
When looking up the original site on [archive.org](https://web.archive.org/web/20150108071555/http://www.csse.monash.edu.au/~jwb/edict.html), I get the following:
```html
<H1>New Edict Home Page</H1>
</center>

The Edict Home Page is now: &nbsp; &nbsp; &nbsp;
<a
href="/web/20150108071555/http://www.edrdg.org/jmdict/edict.html">http://www.edrdg.org/jmdict/edict.html</a>
<p>
Please update your bookmarks.
<p>
You will be transferred there in a few seconds
```

All the other mirrors are still up:
```html
Mirror Sites: &nbsp;
<A HREF="/web/20150109182436/http://www.csse.monash.edu.au/~jwb/cgi-bin/wwwjdic.cgi?1C">Australia</a> &nbsp;
<a href="/web/20150109182436/http://www.ottix.net/cgi-bin/wwwjdic/wwwjdic?1C">Canada</a> &nbsp;
<a href="/web/20150109182436/http://wwwjdic.biz/cgi-bin/wwwjdic?1C">Germany</a> &nbsp;
<a href="/web/20150109182436/http://gengo.com/wwwjdic/cgi-data/wwwjdic?1C">Japan</a> &nbsp;
<a href="/web/20150109182436/http://wwwjdic.se/cgi-bin/wwwjdic.cgi?1C">Sweden</a> &nbsp;
<a href="/web/20150109182436/http://www.edrdg.org/cgi-bin/wwwjdic/wwwjdic?1C">USA</a> 
<a href="/web/20150109182436/http://www.edrdg.org/wwwjdic/wwwjdicmirrors.html">Full List</a> 
```

So I would suggest to move to the American mirror until the Australian site is up again (if ever).
